r_code/placebo_creation.R

```diff
@@ -0,0 +1,217 @@
+    id_cols = c(!!as.name(id_var)),
+    values_from = c(!!as.name(outcome_var))
+  )
+  
+  # Store treated data in a tibble
+  treated_to_match <- data_full %>%
+    dplyr::filter(!!as.name(id_var) %in% tr_entries)
+  
+  #Create a variable indicating the time treatment is assigned by unit
+  treated_to_match <- treated_to_match %>%
+    dplyr::group_by(!!as.name(id_var)) %>%
+    dplyr::mutate(
+      Treatment_Period =
+        length(!!as.name(treat_indicator)) - sum(!!as.name(treat_indicator)) + 1
+    ) %>%
+    dplyr::ungroup()
+  
+  # for pivoting, potential issues arise if we have several time varying 
+  # covariates -- added as TODO.
+  # we'd have to take the values_from each of them, 
+  #and for any constant args we'd presumably have to add them to id_cols
+  #Pivot the data wide, so each row has all the time series data for 
+  #a given unit.
+  data_wide_m <- tidyr::pivot_wider(
+    data = treated_to_match,
+    names_from = !!as.name(time_var),
+    id_cols = c(!!as.name(id_var), Treatment_Period),
+    values_from = c(!!as.name(outcome_var))
+  ) # %>% as.matrix()
+  
+  # Determine control units to assign to placebo treatment group via matching.
+  matched_placebo_df_temp <- matching_without_replacement(data_wide_m, 
+                                                          cd_for_match, 
+                                                          id_var,
+                                                          "Treatment_Period")
+  
+  # We now have the set of control units that form the placebo set, along with 
+  #their placebo Treat Period and their corresponding Treat Unit
+  # Assign treatment to our donor only tibble based on matches.
+  placebo_df_wide <- cd_for_match %>%
+    dplyr::left_join(matched_placebo_df_temp, by = id_var)
+  
+  # Pivot_long the tibble and recreate the treat_indicator indicator
+  # equal to 1 if unit i in time t is treated, 
+  # so that the output matches the input.
+  # If there are covariates or other variables which we do not want to be 
+  # pivoting, append them to non_time_vars vectors
+  # may have to add values_to as well if we have multiple time varying 
+  # variables (added as TODO)
+  non_time_vars <- c(id_var, "Treatment_Period", "Treatment_Unit")
+  placebo_df_long <- placebo_df_wide %>%
+    tidyr::pivot_longer(-all_of(non_time_vars),
+                 names_to = time_var, values_to = outcome_var
+    ) %>%
+    dplyr::mutate(!!as.name(time_var) := as.numeric(!!as.name(time_var))) %>%
+    dplyr::arrange(!!as.name(time_var), !!as.name(id_var)) %>%
+    dplyr::mutate(!!as.name(treat_indicator) := dplyr::case_when(
+      is.na(Treatment_Period) ~ 0,
+      !!as.name(time_var) < Treatment_Period ~ 0,
+      !!as.name(time_var) >= Treatment_Period ~ 1
+    )) %>%
+    dplyr::left_join(
+      data_full %>%
+        dplyr::select(tidyselect::all_of(c(id_var, time_var, counterfac_var))),
+      by = c(id_var, time_var)
+    )
+  
+  return(placebo_df_long)
```

Comment:
The comments were helpful, but I admit I did not fully follow the algorithm.  I will trust it is correct.  No action is needed.  Thanks.